wasm/src/main/scala/ir2wasm/Preprocessor.scala

```diff
@@ -48,4 +64,33 @@ object Preprocessor {
       isAbstract = method.body.isEmpty
     )
   }
+
+  private def collectAbstractMethodCalls(clazz: LinkedClass)(implicit ctx: WasmContext): Unit = {
```

Comment:
I think we don't need to collect abstract methods here.

Preprocessor collects methods' name and signatures in order to later construct vtable (and itables) for the class, and when constructing those tables we collect all the methods defined or declared in super classes and interfaces.

For example,

```scala
class A extends B {
  def a = 1
  override def b: Int = 100
}

class B extends C {
  def b: Int = 1
  override def c: Int = 1
}

abstract class C {
  def c: Int
}
```

`WasmFunctionInfo` for `A`, `B`, `C` are `List(A.a, A.c)`, `List(B.b, B.c)`, `List(C.c)` respectively. (where `C.c` is abstract).

We construct the virtual method tables by correcting all method implementations (`WasmFunctionInfo` where `isAbstract = false`) from the class itself, its superclasses, and interfaces. We remove duplications based on method names, where methods defined in subclasses take precedence. Therefore, the vtable for class A will be `B.c, A.a, A.b`. Actually the global vtable for class A appears as follows:"

```wasm
  (global $sample.A___g_vtable (ref null $sample.A___vtable)
     ref.func $sample.B#c_I___fun
     ref.func $sample.A#a_I___fun
     ref.func $sample.A#b_I___fun
     struct.new $sample.A___vtable
    )
```

And virtual method call `ir.Trees.Apply` (where `receiver` type isn't interface) will be translated to like

```wasm
local.get $instance_of_A ;; receiver for call_ref
;; push args to the stack
local.get $instance_of_A ;; for getting a vtable
struct.get $A 0
struct.get $A_vtable 1 ;; get a function reference to call. the index will be calculated at compile time
call_ref $func_ty
```

Reply:
I initially tried not to do it, but it failed around here:
https://github.com/tanishiking/scala-wasm/blob/6c76da65c24fbe02a3694f6e3218d84d224429dd/wasm/src/main/scala/ir2wasm/WasmExpressionBuilder.scala#L189
because it could not find the period at all.

The problem it's that the frontend linker gets rid of `C.c` entirely. It keeps `B.c` because it's concrete and used. But because `C.c` isn't there at all anymore, if we have `val x: C` and we call `x.c`, we don't find the method at all. You are right that when we have `x: A`, it works without doing this preprocessing.